<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <title>Domain</title>
    <script data-main="js/setup" src="js/require.js"></script>
    <link rel="stylesheet" href="fonts/fonts.css" />
    <link rel="stylesheet" href="fonts/fontawesome-all.min.css" />
    <link rel="stylesheet" href="css/layout.css" />
    <link rel="stylesheet" href="toc/themes/default/style.min.css" />
    <link rel="stylesheet" href="css/design.css" />
  </head>
  <body>

  <div id="control-panel">
    <div id="control-strip-m">
      <div id="control-strip-m-find"><div id="m-find"></div></div>
      <div id="control-strip-m-toc"><div id="m-toc"></div></div>
    </div>
    <div id="controls">
      <div id="control-find">
        <div id="control-find-content"><input id="find" class="typeahead" placeholder="Find something..."></div>
      </div>
      <div id="control-toc">
        <div id="control-toc-content"><div id="toc"></div></div>
      </div>
    </div>
  </div>
  <div id="control-strip-w">
    <div id="control-strip-w-find"></div>
    <div id="control-strip-w-toc"></div>
  </div>

  <div id="content">

  
<div class="heading">
<h1>
  <a name="module_domain" id="module_domain"></a>
  <div class="permalink"><a href="#module_domain" class="permalink fas fa-link"></a></div>
  <p>Domain</p>

</h1>
</div>

<p><strong>This module is pending deprecation</strong>. Once a replacement API has been
finalized, this module will be fully deprecated. Most end users should
<strong>not</strong> have cause to use this module. Users who absolutely must have
the functionality that domains provide may rely on it for the time being
but should expect to have to migrate to a different solution
in the future.</p>
<p>Domains provide a way to handle multiple different IO operations as a
single group.  If any of the event emitters or callbacks registered to a
domain emit an <code>&#39;error&#39;</code> event, or throw an error, then the domain object
will be notified, rather than losing the context of the error in the
<code>process.on(&#39;uncaughtException&#39;)</code> handler, or causing the program to
exit immediately with an error code.</p>


<div class="heading">
<h2>
  <a name="misc_Warning__Don_t_Ignore_Errors_" id="misc_Warning__Don_t_Ignore_Errors_"></a>
  <div class="permalink"><a href="#misc_Warning__Don_t_Ignore_Errors_" class="permalink fas fa-link"></a></div>
  <p>Warning: Don&#39;t Ignore Errors!</p>

</h2>
</div>

<p><p>Domain error handlers are not a substitute for closing down a
process when an error occurs.</p></p>
<p><p>By the very nature of how <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/throw"><code>throw</code></a> works in JavaScript, there is almost
never any way to safely &quot;pick up where you left off&quot;, without leaking
references, or creating some other sort of undefined brittle state.</p></p>
<p><p>The safest way to respond to a thrown error is to shut down the
process. Of course, in a normal web server, there may be many
open connections, and it is not reasonable to abruptly shut those down
because an error was triggered by someone else.</p></p>
<p><p>The better approach is to send an error response to the request that
triggered the error, while letting the others finish in their normal
time, and stop listening for new requests in that worker.</p></p>
<p><p>In this way, <code>domain</code> usage goes hand-in-hand with the cluster module,
since the master process can fork a new worker when a worker
encounters an error.  For Node.js programs that scale to multiple
machines, the terminating proxy or service registry can take note of
the failure, and react accordingly.</p></p>
<p><p>For example, this is not a good idea:</p></p>
<pre class="prettyprint"><code class="lang-js">// XXX WARNING!  BAD IDEA!

const d = require(&#39;domain&#39;).create();
d.on(&#39;error&#39;, (er) =&gt; {
  // The error won&#39;t crash the process, but what it does is worse!
  // Though we&#39;ve prevented abrupt process restarting, we are leaking
  // resources like crazy if this ever happens.
  // This is no better than process.on(&#39;uncaughtException&#39;)!
  console.log(`error, but oh well ${er.message}`);
});
d.run(() =&gt; {
  require(&#39;http&#39;).createServer((req, res) =&gt; {
    handleRequest(req, res);
  }).listen(PORT);
});
</code></pre>
<p>By using the context of a domain, and the resilience of separating our
program into multiple worker processes, we can react more
appropriately, and handle errors with much greater safety.</p>
<pre class="prettyprint"><code class="lang-js">// Much better!

const cluster = require(&#39;cluster&#39;);
const PORT = +process.env.PORT || 1337;

if (cluster.isMaster) {
  // A more realistic scenario would have more than 2 workers,
  // and perhaps not put the master and worker in the same file.
  //
  // It is also possible to get a bit fancier about logging, and
  // implement whatever custom logic is needed to prevent DoS
  // attacks and other bad behavior.
  //
  // See the options in the cluster documentation.
  //
  // The important thing is that the master does very little,
  // increasing our resilience to unexpected errors.

  cluster.fork();
  cluster.fork();

  cluster.on(&#39;disconnect&#39;, (worker) =&gt; {
    console.error(&#39;disconnect!&#39;);
    cluster.fork();
  });

} else {
  // the worker
  //
  // This is where we put our bugs!

  const domain = require(&#39;domain&#39;);

  // See the cluster documentation for more details about using
  // worker processes to serve requests.  How it works, caveats, etc.

  const server = require(&#39;http&#39;).createServer((req, res) =&gt; {
    const d = domain.create();
    d.on(&#39;error&#39;, (er) =&gt; {
      console.error(`error ${er.stack}`);

      // Note: We&#39;re in dangerous territory!
      // By definition, something unexpected occurred,
      // which we probably didn&#39;t want.
      // Anything can happen now!  Be very careful!

      try {
        // make sure we close down within 30 seconds
        const killtimer = setTimeout(() =&gt; {
          process.exit(1);
        }, 30000);
        // But don&#39;t keep the process open just for that!
        killtimer.unref();

        // stop taking new requests.
        server.close();

        // Let the master know we&#39;re dead.  This will trigger a
        // &#39;disconnect&#39; in the cluster master, and then it will fork
        // a new worker.
        cluster.worker.disconnect();

        // try to send an error to the request that triggered the problem
        res.statusCode = 500;
        res.setHeader(&#39;content-type&#39;, &#39;text/plain&#39;);
        res.end(&#39;Oops, there was a problem!\n&#39;);
      } catch (er2) {
        // oh well, not much we can do at this point.
        console.error(`Error sending 500! ${er2.stack}`);
      }
    });

    // Because req and res were created before this domain existed,
    // we need to explicitly add them.
    // See the explanation of implicit vs explicit binding below.
    d.add(req);
    d.add(res);

    // Now run the handler function in the domain.
    d.run(() =&gt; {
      handleRequest(req, res);
    });
  });
  server.listen(PORT);
}

// This part is not important.  Just an example routing thing.
// Put fancy application logic here.
function handleRequest(req, res) {
  switch (req.url) {
    case &#39;/error&#39;:
      // We do some async stuff, and then...
      setTimeout(() =&gt; {
        // Whoops!
        flerb.bark();
      }, timeout);
      break;
    default:
      res.end(&#39;ok&#39;);
  }
}
</code></pre>


<div class="heading">
<h2>
  <a name="misc_Additions_to_Error_objects" id="misc_Additions_to_Error_objects"></a>
  <div class="permalink"><a href="#misc_Additions_to_Error_objects" class="permalink fas fa-link"></a></div>
  <p>Additions to Error objects</p>

</h2>
</div>

<p><p>Any time an <code>Error</code> object is routed through a domain, a few extra fields
are added to it.</p></p>
<ul>
<li><code>error.domain</code> The domain that first handled the error.</li>
<li><code>error.domainEmitter</code> The event emitter that emitted an <code>&#39;error&#39;</code> event
with the error object.</li>
<li><code>error.domainBound</code> The callback function which was bound to the
domain, and passed an error as its first argument.</li>
<li><code>error.domainThrown</code> A boolean indicating whether the error was
thrown, emitted, or passed to a bound callback function.</li>
</ul>


<div class="heading">
<h2>
  <a name="misc_Implicit_Binding" id="misc_Implicit_Binding"></a>
  <div class="permalink"><a href="#misc_Implicit_Binding" class="permalink fas fa-link"></a></div>
  <p>Implicit Binding</p>

</h2>
</div>

<p>If domains are in use, then all <strong>new</strong> EventEmitter objects (including
Stream objects, requests, responses, etc.) will be implicitly bound to
the active domain at the time of their creation.</p>
<p>Additionally, callbacks passed to lowlevel event loop requests (such as
to fs.open, or other callback-taking methods) will automatically be
bound to the active domain. If they throw, then the domain will catch
the error.</p>
<p>In order to prevent excessive memory usage, Domain objects themselves
are not implicitly added as children of the active domain.  If they
were, then it would be too easy to prevent request and response objects
from being properly garbage collected.</p>
<p>To nest Domain objects as children of a parent Domain they must be explicitly
added.</p>
<p>Implicit binding routes thrown errors and <code>&#39;error&#39;</code> events to the
Domain&#39;s <code>&#39;error&#39;</code> event, but does not register the EventEmitter on the
Domain.
Implicit binding only takes care of thrown errors and <code>&#39;error&#39;</code> events.</p>


<div class="heading">
<h2>
  <a name="misc_Explicit_Binding" id="misc_Explicit_Binding"></a>
  <div class="permalink"><a href="#misc_Explicit_Binding" class="permalink fas fa-link"></a></div>
  <p>Explicit Binding</p>

</h2>
</div>

<p><p>Sometimes, the domain in use is not the one that ought to be used for a
specific event emitter.  Or, the event emitter could have been created
in the context of one domain, but ought to instead be bound to some
other domain.</p></p>
<p><p>For example, there could be one domain in use for an HTTP server, but
perhaps we would like to have a separate domain to use for each request.</p></p>
<p><p>That is possible via explicit binding.</p></p>
<p><p>For example:</p></p>
<pre class="prettyprint"><code class="lang-js">// create a top-level domain for the server
const domain = require(&#39;domain&#39;);
const http = require(&#39;http&#39;);
const serverDomain = domain.create();

serverDomain.run(() =&gt; {
  // server is created in the scope of serverDomain
  http.createServer((req, res) =&gt; {
    // req and res are also created in the scope of serverDomain
    // however, we&#39;d prefer to have a separate domain for each request.
    // create it first thing, and add req and res to it.
    const reqd = domain.create();
    reqd.add(req);
    reqd.add(res);
    reqd.on(&#39;error&#39;, (er) =&gt; {
      console.error(&#39;Error&#39;, er, req.url);
      try {
        res.writeHead(500);
        res.end(&#39;Error occurred, sorry.&#39;);
      } catch (er2) {
        console.error(&#39;Error sending 500&#39;, er2, req.url);
      }
    });
  }).listen(1337);
});
</code></pre>


<div class="heading">
<h2>
  <a name="module_domains_and_promises" id="module_domains_and_promises"></a>
  <div class="permalink"><a href="#module_domains_and_promises" class="permalink fas fa-link"></a></div>
  <p>Domains and Promises</p>

</h2>
</div>

<p><p>As of Node 8.0.0, the handlers of Promises are run inside the domain in
which the call to <code>.then</code> or <code>.catch</code> itself was made:</p></p>
<p><pre class="prettyprint"><code class="lang-js">const d1 = domain.create();
const d2 = domain.create();</p>
<p>let p;
d1.run(() =&gt; {
  p = Promise.resolve(42);
});</p>
<p>d2.run(() =&gt; {
  p.then((v) =&gt; {
    // running in d2
  });
});
</code></pre></p>
<p><p>A callback may be bound to a specific domain using <a href="#domain_domain_bind_callback"><code>domain.bind(callback)</code></a>:</p></p>
<p><pre class="prettyprint"><code class="lang-js">const d1 = domain.create();
const d2 = domain.create();</p>
<p>let p;
d1.run(() =&gt; {
  p = Promise.resolve(42);
});</p>
<p>d2.run(() =&gt; {
  p.then(p.domain.bind((v) =&gt; {
    // running in d1
  }));
});
</code></pre></p>
<p><p>Note that domains will not interfere with the error handling mechanisms for
Promises, i.e. no <code>error</code> event will be emitted for unhandled Promise
rejections.</p>
<!-- [end-include:domain.md] -->
<!-- [start-include:esm.md] --></p>

<div class='expandable-header' id='class_Domain-expand'>
<div class="heading">
    <h2>
        <a name="class_Domain" id="class_Domain"></a>
        <i class='expandable-icon fas fa-caret-right' id="class_Domain-expand-icon"></i>
        <div class="permalink"><a href="#class_Domain" class="permalink fas fa-link"></a></div>
        <p>Class: Domain</p>

    </h2>
</div>

</div>
<div class='expandable-content' id='class_Domain-expand-content'>
<p>The Domain class encapsulates the functionality of routing errors and
uncaught exceptions to the active Domain object.</p>
<p>Domain is a child class of <a href="events.html"><code>EventEmitter</code></a>.  To handle the errors that it
catches, listen to its <code>&#39;error&#39;</code> event.</p>


<div class="heading">
<h3>
  <a name="subgroup_class_Domain_Properties" id="subgroup_class_Domain_Properties"></a>
  <div class="permalink"><a href="#subgroup_class_Domain_Properties" class="permalink fas fa-link"></a></div>
  <p>Properties</p>

</h3>
</div>


<div class='expandable-header' id='property_members-expand'>
<div class="heading">
    <h4>
        <a name="property_members" id="property_members"></a>
        <i class='expandable-icon fas fa-caret-right' id="property_members-expand-icon"></i>
        <div class="permalink"><a href="#property_members" class="permalink fas fa-link"></a></div>
        <p><code>members</code> {Array} </p>

    </h4>
</div>

</div>
<div class='expandable-content' id='property_members-expand-content'>
<p>An array of timers and event emitters that have been explicitly added
to the domain.</p>


</div>
<div class="heading">
<h3>
  <a name="subgroup_class_Domain_Methods" id="subgroup_class_Domain_Methods"></a>
  <div class="permalink"><a href="#subgroup_class_Domain_Methods" class="permalink fas fa-link"></a></div>
  <p>Methods</p>

</h3>
</div>


<div class='expandable-header' id='method_add-expand'>
<div class="heading">
    <h4>
        <a name="method_add" id="method_add"></a>
        <i class='expandable-icon fas fa-caret-right' id="method_add-expand-icon"></i>
        <div class="permalink"><a href="#method_add" class="permalink fas fa-link"></a></div>
        <p>domain.add(emitter)</p>

    </h4>
</div>

</div>
<div class='expandable-content' id='method_add-expand-content'>
<p>Explicitly adds an emitter to the domain.  If any event handlers called by
the emitter throw an error, or if the emitter emits an <code>&#39;error&#39;</code> event, it
will be routed to the domain&#39;s <code>&#39;error&#39;</code> event, just like with implicit
binding.</p>
<p>This also works with timers that are returned from <a href="timers.html#timers_setinterval_callback_delay_args"><code>setInterval()</code></a> and
<a href="timers.html#timers_settimeout_callback_delay_args"><code>setTimeout()</code></a>.  If their callback function throws, it will be caught by
the domain &#39;error&#39; handler.</p>
<p>If the Timer or EventEmitter was already bound to a domain, it is removed
from that one, and bound to this one instead.</p>


</div><div class='expandable-header' id='method_bind-expand'>
<div class="heading">
    <h4>
        <a name="method_bind" id="method_bind"></a>
        <i class='expandable-icon fas fa-caret-right' id="method_bind-expand-icon"></i>
        <div class="permalink"><a href="#method_bind" class="permalink fas fa-link"></a></div>
        <p>domain.bind(callback)</p>

    </h4>
</div>

</div>
<div class='expandable-content' id='method_bind-expand-content'>
<p><p>The returned function will be a wrapper around the supplied callback
function.  When the returned function is called, any errors that are
thrown will be routed to the domain&#39;s <code>&#39;error&#39;</code> event.</p></p>
<p><h4>Example</h4></p>
<pre class="prettyprint"><code class="lang-js">const d = domain.create();

function readSomeFile(filename, cb) {
  fs.readFile(filename, &#39;utf8&#39;, d.bind((er, data) =&gt; {
    // if this throws, it will also be passed to the domain
    return cb(er, data ? JSON.parse(data) : null);
  }));
}

d.on(&#39;error&#39;, (er) =&gt; {
  // an error occurred somewhere.
  // if we throw it now, it will crash the program
  // with the normal line number and stack message.
});
</code></pre>


</div><div class='expandable-header' id='method_enter-expand'>
<div class="heading">
    <h4>
        <a name="method_enter" id="method_enter"></a>
        <i class='expandable-icon fas fa-caret-right' id="method_enter-expand-icon"></i>
        <div class="permalink"><a href="#method_enter" class="permalink fas fa-link"></a></div>
        <p>domain.enter()</p>

    </h4>
</div>

</div>
<div class='expandable-content' id='method_enter-expand-content'>
<p>The <code>enter</code> method is plumbing used by the <code>run</code>, <code>bind</code>, and <code>intercept</code>
methods to set the active domain. It sets <code>domain.active</code> and <code>process.domain</code>
to the domain, and implicitly pushes the domain onto the domain stack managed
by the domain module (see <a href="#domain_domain_exit"><code>domain.exit()</code></a> for details on the domain stack). The
call to <code>enter</code> delimits the beginning of a chain of asynchronous calls and I/O
operations bound to a domain.</p>
<p>Calling <code>enter</code> changes only the active domain, and does not alter the domain
itself. <code>enter</code> and <code>exit</code> can be called an arbitrary number of times on a
single domain.</p>


</div><div class='expandable-header' id='method_exit_1-expand'>
<div class="heading">
    <h4>
        <a name="method_exit_1" id="method_exit_1"></a>
        <i class='expandable-icon fas fa-caret-right' id="method_exit_1-expand-icon"></i>
        <div class="permalink"><a href="#method_exit_1" class="permalink fas fa-link"></a></div>
        <p>domain.exit()</p>

    </h4>
</div>

</div>
<div class='expandable-content' id='method_exit_1-expand-content'>
<p>The <code>exit</code> method exits the current domain, popping it off the domain stack.
Any time execution is going to switch to the context of a different chain of
asynchronous calls, it&#39;s important to ensure that the current domain is exited.
The call to <code>exit</code> delimits either the end of or an interruption to the chain
of asynchronous calls and I/O operations bound to a domain.</p>
<p>If there are multiple, nested domains bound to the current execution context,
<code>exit</code> will exit any domains nested within this domain.</p>
<p>Calling <code>exit</code> changes only the active domain, and does not alter the domain
itself. <code>enter</code> and <code>exit</code> can be called an arbitrary number of times on a
single domain.</p>


</div><div class='expandable-header' id='method_intercept-expand'>
<div class="heading">
    <h4>
        <a name="method_intercept" id="method_intercept"></a>
        <i class='expandable-icon fas fa-caret-right' id="method_intercept-expand-icon"></i>
        <div class="permalink"><a href="#method_intercept" class="permalink fas fa-link"></a></div>
        <p>domain.intercept(callback)</p>

    </h4>
</div>

</div>
<div class='expandable-content' id='method_intercept-expand-content'>
<p><p>This method is almost identical to <a href="#domain_domain_bind_callback"><code>domain.bind(callback)</code></a>.  However, in
addition to catching thrown errors, it will also intercept <a href="errors.html#errors_class_error"><code>Error</code></a>
objects sent as the first argument to the function.</p></p>
<p><p>In this way, the common <code>if (err) return callback(err);</code> pattern can be replaced
with a single error handler in a single place.</p></p>
<p><h4>Example</h4></p>
<pre class="prettyprint"><code class="lang-js">const d = domain.create();

function readSomeFile(filename, cb) {
  fs.readFile(filename, &#39;utf8&#39;, d.intercept((data) =&gt; {
    // note, the first argument is never passed to the
    // callback since it is assumed to be the &#39;Error&#39; argument
    // and thus intercepted by the domain.

    // if this throws, it will also be passed to the domain
    // so the error-handling logic can be moved to the &#39;error&#39;
    // event on the domain instead of being repeated throughout
    // the program.
    return cb(null, JSON.parse(data));
  }));
}

d.on(&#39;error&#39;, (er) =&gt; {
  // an error occurred somewhere.
  // if we throw it now, it will crash the program
  // with the normal line number and stack message.
});
</code></pre>


</div><div class='expandable-header' id='method_remove-expand'>
<div class="heading">
    <h4>
        <a name="method_remove" id="method_remove"></a>
        <i class='expandable-icon fas fa-caret-right' id="method_remove-expand-icon"></i>
        <div class="permalink"><a href="#method_remove" class="permalink fas fa-link"></a></div>
        <p>domain.remove(emitter)</p>

    </h4>
</div>

</div>
<div class='expandable-content' id='method_remove-expand-content'>
<p>The opposite of <a href="#domain_domain_add_emitter"><code>domain.add(emitter)</code></a>.  Removes domain handling from the
specified emitter.</p>


</div><div class='expandable-header' id='method_run-expand'>
<div class="heading">
    <h4>
        <a name="method_run" id="method_run"></a>
        <i class='expandable-icon fas fa-caret-right' id="method_run-expand-icon"></i>
        <div class="permalink"><a href="#method_run" class="permalink fas fa-link"></a></div>
        <p>domain.run(fn[, ...args])</p>

    </h4>
</div>

</div>
<div class='expandable-content' id='method_run-expand-content'>
<p>Run the supplied function in the context of the domain, implicitly
binding all event emitters, timers, and lowlevel requests that are
created in that context. Optionally, arguments can be passed to
the function.</p>
<p>This is the most basic way to use a domain.</p>
<p>Example:</p>
<pre class="prettyprint"><code class="lang-js">const domain = require(&#39;domain&#39;);
const fs = require(&#39;fs&#39;);
const d = domain.create();
d.on(&#39;error&#39;, (er) =&gt; {
  console.error(&#39;Caught error!&#39;, er);
});
d.run(() =&gt; {
  process.nextTick(() =&gt; {
    setTimeout(() =&gt; { // simulating some various async stuff
      fs.open(&#39;non-existent file&#39;, &#39;r&#39;, (er, fd) =&gt; {
        if (er) throw er;
        // proceed...
      });
    }, 100);
  });
});
</code></pre>
<p>In this example, the <code>d.on(&#39;error&#39;)</code> handler will be triggered, rather
than crashing the program.</p>


</div>
</div><div class='expandable-header' id='method_create-expand'>
<div class="heading">
    <h2>
        <a name="method_create" id="method_create"></a>
        <i class='expandable-icon fas fa-caret-right' id="method_create-expand-icon"></i>
        <div class="permalink"><a href="#method_create" class="permalink fas fa-link"></a></div>
        <p>domain.create()</p>

    </h2>
</div>

</div>
<div class='expandable-content' id='method_create-expand-content'>
<p>Returns a new Domain object.</p>


</div>

  <div class="footer">
    <p>Some footer text. Optional. Copyright &copy; whenever by your bidness.</p>
    <p>Dubbed by <img src="images/mixtape.png" title="Dubbed by mixtape." style="width:40px;position:relative;top:8px"></img> mixtape.</p>
  </div>

  </div>
  </body>
</html>